libudpard_demo/src/main.c

```diff
@@ -0,0 +1,1546 @@
+            rx_aw[rx_count].user_reference = &app->sub_data;
+            rx_count++;
+        }
+        assert(rx_count <= (sizeof(rx_aw) / sizeof(rx_aw[0])));
+    }
+    if (app->local_node_id <= UDPARD_NODE_ID_MAX)  // The RPC socket is not initialized until the node-ID is set.
+    {
+        for (size_t i = 0; i < app->iface_count; i++)  // RPC dispatcher sockets (one per iface).
+        {
+            rx_aw[rx_count].handle         = &app->rpc_dispatcher.io[i];
+            rx_aw[rx_count].user_reference = NULL;
+            rx_count++;
+            assert(rx_count <= (sizeof(rx_aw) / sizeof(rx_aw[0])));
+        }
+    }
+
+    // Block until something happens or the deadline is reached.
+    const int16_t wait_result = udpWait((unblock_deadline > ts_before_usec) ? (unblock_deadline - ts_before_usec) : 0,
+                                        tx_count,
+                                        &tx_aw[0],
+                                        rx_count,
+                                        &rx_aw[0]);
+    if (wait_result < 0)
+    {
+        abort();  // Unreachable.
+    }
+
+    // Process the RX sockets that became readable.
+    // The time has to be re-sampled because the blocking wait may have taken a long time.
+    const UdpardMicrosecond ts_after_usec = getMonotonicMicroseconds();
+    for (size_t i = 0; i < rx_count; i++)
+    {
+        if (!rx_aw[i].ready)
+        {
+            continue;  // This one is not yet ready to read.
+        }
+        // Allocate memory that we will read the data into. The ownership of this memory will be transferred
+        // to LibUDPard, which will free it when it is no longer needed.
+        // A deeply embedded system may be able to transfer this memory directly from the NIC driver to eliminate copy.
+        struct UdpardMutablePayload payload = {
+            .size = RX_BUFFER_SIZE,
+            .data = app->memory.payload.allocate(app->memory.payload.user_reference, RX_BUFFER_SIZE),
+        };
+        if (NULL == payload.data)
+        {
+            (void) fprintf(stderr, "RX payload allocation failure: out of memory\n");
+            continue;
+        }
+        // Read the data from the socket into the buffer we just allocated.
+        const int16_t rx_result = udpRxReceive(rx_aw[i].handle, &payload.size, payload.data);
+        assert(0 != rx_result);
+        if (rx_result < 0)
+        {
+            // We end up here if the socket was closed while processing another datagram.
+            // This happens if a subscriber chose to unsubscribe dynamically.
+            app->memory.payload.deallocate(app->memory.payload.user_reference, RX_BUFFER_SIZE, payload.data);
+            continue;
+        }
+        // Pass the data buffer into LibUDPard for further processing. It takes ownership of the buffer.
+        if (rx_aw[i].user_reference != NULL)  // This is used to differentiate subscription sockets from RPC sockets.
```

Comment:
I ask myself the same question. We need to keep state per awaitable to differentiate between the two. We can't subtype it to add custom state because that breaks the array layout. All of the better solutions I could come up with are various shades of bad:

1. Pass awaitables as an array of pointers -- requires an extra array.
2. Use a linked list -- results in a clumsy API.
3. Add the required fields to the awaitable type -- breaks encapsulation.

I guess I better add an explanatory comment about this.